src/main/java/io/numaproj/numaflow/function/FunctionService.java

```diff
@@ -154,59 +152,56 @@ public StreamObserver<Udfunction.Datum> reduceFn(StreamObserver<Udfunction.Datum
-                                datum.getWatermark().getWatermark().getSeconds(),
-                                datum.getWatermark().getWatermark().getNanos()),
-                        Instant.ofEpochSecond(
-                                datum.getEventTime().getEventTime().getSeconds(),
-                                datum.getEventTime().getEventTime().getNanos()),
-                        false);
-                try {
-                    if (!streamMap.containsKey(datum.getKey())) {
-                        ReduceDatumStreamImpl reduceDatumStream = new ReduceDatumStreamImpl();
-                        results.add(reduceTaskExecutor.submit(() -> reduceHandler.HandleDo(
-                                datum.getKey(),
-                                reduceDatumStream,
-                                md)));
-                        streamMap.put(datum.getKey(), reduceDatumStream);
-                    }
-                    streamMap.get(datum.getKey()).WriteMessage(handlerDatum);
-                } catch (InterruptedException e) {
-                    Thread.interrupted();
-                    onError(e);
+                // send the message to parent actor, which takes care of distribution.
+                if (!parentActorRef.isTerminated()) {
+                    parentActorRef.tell(datum, parentActorRef);
+                } else {
+                    responseObserver.onError(new Throwable("Supervisor actor was terminated"));
                 }
             }
 
             @Override
             public void onError(Throwable throwable) {
-                logger.log(Level.WARNING, "Encountered error in reduceFn", throwable);
+                log.error("error from the client" + throwable.getMessage());
                 responseObserver.onError(throwable);
             }
 
             @Override
             public void onCompleted() {
-                Udfunction.DatumList response = Udfunction.DatumList
-                        .newBuilder()
-                        .getDefaultInstanceForType();
+
+                // Ask the parent to return the list of futures returned by child actors.
+                Future<Object> resultFuture = Patterns.ask(parentActorRef, EOF, Integer.MAX_VALUE);
+
+                List<Future<Object>> udfResultFutures;
                 try {
-                    for (Map.Entry<String, ReduceDatumStreamImpl> entry : streamMap.entrySet()) {
-                        entry.getValue().WriteMessage(ReduceDatumStream.EOF);
-                    }
-                    response = buildDatumListResponse(results);
-                } catch (InterruptedException | ExecutionException e) {
-                    Thread.interrupted();
-                    onError(e);
+                    udfResultFutures = (List<Future<Object>>) Await.result(
+                            resultFuture,
+                            Duration.Inf());
+                } catch (TimeoutException | InterruptedException e) {
```

Comment:
editor doesn't allow me to remove that :)

Reply:
Right, we can't control how long it takes for child actors to return a future. But I still think the system should set a proper timeout to avoid being blocked forever. something like http client request timeout, if server doesn't reply within x seconds, we consider the request as failed and handle it from client side.

I am not sure about the actual timeout value we should set atm. Maybe we can push this out and monitor it to determine.